src/query/procedure/py_module.cpp

```diff
@@ -905,30 +918,66 @@ std::optional<py::ExceptionInfo> AddRecordFromPython(mgp_result *result, py::Obj
       PyErr_SetString(PyExc_TypeError, msg.c_str());
       return py::FetchError();
     }
-    const auto *field_name = PyUnicode_AsUTF8(key);
+    const char *field_name = PyUnicode_AsUTF8(key);
     if (!field_name) return py::FetchError();
-    auto *val = PyTuple_GetItem(item, 1);
+    PyObject *val = PyTuple_GetItem(item, 1);
     if (!val) return py::FetchError();
     // This memory is one dedicated for mg_procedure.
     mgp_value *field_val = PyObjectToMgpValueWithPythonExceptions(val, memory);
     if (field_val == nullptr) {
       return py::FetchError();
     }
-    if (mgp_result_record_insert(record, field_name, field_val) != mgp_error::MGP_ERROR_NO_ERROR) {
+
+    if (!is_transactional) {
+      if (ContainsDeleted(field_val)) {
+        mgp_value_destroy(field_val);
+        for (auto &cache_entry : current_record_cache) {
+          mgp_value_destroy(cache_entry.field_val);
+        }
+
+        return std::nullopt;
+      }
+      current_record_cache.emplace_back(
+          RecordFieldCache{.key = key, .val = val, .field_name = field_name, .field_val = field_val});
+    } else {
+      if (mgp_result_record_insert(record, field_name, field_val) != mgp_error::MGP_ERROR_NO_ERROR) {
+        std::stringstream ss;
+        ss << "Unable to insert field '" << py::Object::FromBorrow(key) << "' with value: '"
+           << py::Object::FromBorrow(val) << "'; did you set the correct field type?";
+        const auto &msg = ss.str();
+        PyErr_SetString(PyExc_ValueError, msg.c_str());
+        mgp_value_destroy(field_val);
+        return py::FetchError();
+      }
+      mgp_value_destroy(field_val);
+    }
+  }
+
+  if (is_transactional) {
+    return std::nullopt;
+  }
+
+  if (RaiseExceptionFromErrorCode(mgp_result_new_record(result, &record))) {
+    return py::FetchError();
+  }
+  for (auto &cache_entry : current_record_cache) {
+    if (mgp_result_record_insert(record, cache_entry.field_name, cache_entry.field_val) !=
+        mgp_error::MGP_ERROR_NO_ERROR) {
       std::stringstream ss;
-      ss << "Unable to insert field '" << py::Object::FromBorrow(key) << "' with value: '"
-         << py::Object::FromBorrow(val) << "'; did you set the correct field type?";
+      ss << "Unable to insert field '" << py::Object::FromBorrow(cache_entry.key) << "' with value: '"
+         << py::Object::FromBorrow(cache_entry.val) << "'; did you set the correct field type?";
       const auto &msg = ss.str();
       PyErr_SetString(PyExc_ValueError, msg.c_str());
-      mgp_value_destroy(field_val);
+      mgp_value_destroy(cache_entry.field_val);
       return py::FetchError();
     }
-    mgp_value_destroy(field_val);
+    mgp_value_destroy(cache_entry.field_val);
```

Comment:
This is repeated code, I would isolate this in lambda

Reply:
If I am not wrong, we can do that

Reply:
Thanks for the suggestion! I’ve isolated this in a function of its own (in order to make the parent function simpler)